USAGE.adoc

```diff
@@ -0,0 +1,255 @@
+
+[source,java]
+----
+ClassCreator creator = ClassCreator.builder() <1>
+                             .classOutput(cl) <2>
+                             .className("com.MyTest") <3>
+                             .build() <4>
+----
+<1> Create a new `ClassCreator.Builder` instance.
+<2> Set the `ClassOutput` that is used to write the bytecode of the class.
+<3> Specify the fully-qualified class name.
+<4> Build the `ClassCreator` instance.
+
+By default, the generated class is `public` and _synthetic_ (as defined by the Java Language Specification).
+If you need to make it `final` then use the `ClassCreator.Builder#setFinal()` method.
+
+The `ClassCreator.Builder#superClass()` and ``ClassCreator.Builder#interfaces()` methods can be used to specify the superclass and implemented interfaces.
+
+== Adding Fields
+
+Once we have a `ClassCreator` instance we can start adding fields and methods.
+Fields can be added via the `FieldCreator` interface returned from the `ClassCreator#getFieldCreator()` methods.
+By default, the generated field is `private`.
+You can use the `FieldCreator#setModifiers()` method to control the modifiers of the generated field.
+
+.`FieldCreator` Example
+[source,java]
+----
+import org.objectweb.asm.Opcodes.ACC_FINAL;
+import org.objectweb.asm.Opcodes.ACC_PROTECTED;
+
+void addField(ClassCreator fooClazz) {
+  FieldCreator myField = fooClazz.getFieldCreator("score", Integer.class); <1>
+  myField.setModifiers(ACC_FINAL | ACC_PROTECTED); <2>
+}
+----
+<1> Creates a field `score` of type `Integer`.
+<2> Makes the field `protected final`.
+
+== Adding Methods
+
+A method can be added through a `MethodCreator` interface returned from the `ClassCreator#getMethodCreator()` methods.
+By default, the generated field is `public`.
+You can use the `MethodCreator#setModifiers()` method to control the modifiers of the generated field.
+
+The bytecode of the method is generated via `BytecodeCreator`.
+See <<method_bytecode>> for more information about how to generate some most common operations.
+
+.`MethodCreator` Example
+[source,java]
+----
+import org.objectweb.asm.Opcodes.ACC_PRIVATE;
+
+void addMethod(ClassCreator fooClazz) {
+  MethodCreator alwaysReturnFalse = fooClazz.getMethodCreator("alwaysReturnFalse", boolean.class); <1>
+  alwaysReturnFalse.setModifiers(ACC_PRIVATE);
+  // Note that MethodCreator also implements BytecodeCreator
+  alwaysReturnFalse.returnValue(alwaysReturnFalse.load(false)); <2> <3>
+}
+----
+<1> Create a method with the following signature `private boolean alwaysReturnFalse()`.
+<2> A `MethodCreator` must always define a return statement.
+<3> This is an equivalent of `return false;`.
+
+[[method_bytecode]]
+== Generating Method Bytecode
+
+The `BytecodeCreator` interface allows you to generate some most common bytecode operations.
+It's not intended to cover all the functionality provided by JVM and Java language though.
+Most operations return and accept a `ResultHandle`.
```

Comment:
I'm not so sure about this description of `ResultHandle` but I didn't find a better words... so feel free to reword my initial attempt.

Reply:
I think your explanation makes sense, even if it is factually incorrect.

(Gizmo allocates a local variable for almost all `ResultHandle`s, with the exception of a top-of-stack value that is a direct input to the immediately following operation, which is kept on stack and not spilled into a local.)